_source/_posts/2021-09-24-java-records.md

```diff
@@ -0,0 +1,716 @@
+            .withExposedPorts(27017)
+            .withCopyFileToContainer(MountableFile.forClasspathResource("stats.json"),
+                    "/stats.json")
+            .withEnv("MONGO_INIT_DATABASE", "fortnite");
+
+    @BeforeAll
+    public static void setUp() throws IOException, InterruptedException {
+        mongoDBContainer.setPortBindings(List.of("27017:27017"));
+        mongoDBContainer.start();
+
+        Container.ExecResult result = mongoDBContainer.execInContainer("mongoimport",
+                "--verbose", "--db=fortnite", "--collection=stats", "--file=/stats.json", "--jsonArray");
+        logger.info(result.getStdout());
+        logger.info(result.getStderr());
+        logger.info("exit code={}", result.getExitCode());
+    }
+
+
+    @Test
+    public void testGetAll(){
+        Flux<EndOfGame> stats = statsRepository.findAll();
+        List<EndOfGame> result = stats.collectList().block();
+
+        assertThat(result).isNotEmpty();
+        assertThat(result).size().isEqualTo(87);
+    }
+
+    @Test
+    public void testQueryMentalStateAverageDamage(){
+        Flux<MentalStateDamage> stats = statsRepository.queryMentalStateAverageDamage();
+        List<MentalStateDamage> result = stats.collectList().block();
+
+        assertThat(result).isNotEmpty();
+        assertThat(result).size().isEqualTo(2);
+        assertThat(result.get(0).mentalState()).isEqualTo("sober");
+        assertThat(result.get(1).mentalState()).isEqualTo("high");
+
+        logger.info(result.get(0).toString());
+        logger.info(result.get(1).toString());
+    }
+
+    @AfterAll
+    public static void tearDown(){
+        mongoDBContainer.stop();
+    }
+
+}
+```
+
+`@DataMongoTest` configures the data layer for testing. For this test, I was not able to configure a [repository populator](https://docs.spring.io/spring-data/mongodb/docs/current/reference/html/#core.repository-populators), so with the help of Testcontainers a MongoDB instance is started instead. Using a container MongoDB instance requires disabling the `EmbeddedMongoAutoConfiguration`.
+
+In the `setUp()` above, the `mongoimport` tool is executed in the test container, initializing the `stats` collection with the sample dataset. If the import runs successfully, the following line should appear in the test logs:
+
+```text
+87 document(s) imported successfully. 0 document(s) failed to import.
+```
+Also, you can inspect the response in the logs for the average damage test:
+
+```
+MentalStateDamage[mentalState=sober, damageToPlayers=604.3777777777777, damageToStructures=3373.511111111111, damageTaken=246.46666666666667]
+MentalStateDamage[mentalState=high, damageToPlayers=557.547619047619, damageToStructures=2953.8571428571427, damageTaken=241.71428571428572]
+```
+For this single-player dataset, the damage taken or inflicted was not orders of magnitude different when sober than when high.
```

Comment:
I love the tests in this section! ❤️ 

Why weren't you able to configure a repository populator? You could just put a positive spin on it and say something like:

> For this test, I decided not to configure a [repository populator]() and used Testcontainers instead.

Reply:
You never mention how to run the tests to see the output. I'd add a sentence that says run `mvn test` or `mvn test -Dtestname`, or tell them to run the test from their IDE.